core/src/main/scala/akka/kafka/internal/ConsumerStage.scala

```diff
@@ -0,0 +1,267 @@
+
+      private val decrementConfirmation = getAsyncCallback[Unit] { _ =>
+        awaitingConfirmation -= 1
+        logger.trace(s"Commits in progress {}", awaitingConfirmation.toString)
+      }
+
+      // impl of GraphStage.Committer that is called from the outside via CommittableOffsetImpl
+      override def commit(offset: Consumer.Offset): Future[Done] = {
+        // committed offset should be the next message the application will consume, i.e. +1
+        commit(Map(new TopicPartition(offset.topic, offset.partition) ->
+          new OffsetAndMetadata(offset.offset + 1)))
+      }
+
+      // impl of GraphStage.Committer that is called from the outside via CommittableOffsetBatchImpl
+      override def commit(offsets: Map[TopicPartition, OffsetAndMetadata]): Future[Done] = {
+        // not important that we complete with failure during stopping race,
+        // it will be completed with TimeoutException
+        if (stoppedPromise.isCompleted)
+          Future.failed(new IllegalStateException("ConsumerStage is stopped, offset commit not performed"))
+        else {
+          val done = Promise[Done]()
+          val c = materializer.scheduleOnce(COMMIT_TIMEOUT, new Runnable {
+            override def run(): Unit = {
+              done.tryFailure(new TimeoutException(
+                s"Offset commit timed out after ${COMMIT_TIMEOUT.toMillis} ms, $offsets"
+              ))
+            }
+          })
+          done.future.onComplete(__ => c.cancel())(materializer.executionContext)
+          commitCallback.invoke((offsets, done))
+          done.future
+        }
+      }
+
+      val commitCallback = getAsyncCallback[(Map[TopicPartition, OffsetAndMetadata], Promise[Done])] {
+        case (offsets, done) => commitInternal(offsets, done)
+      }
+
+      private def commitInternal(offsets: Map[TopicPartition, OffsetAndMetadata], done: Promise[Done]): Unit = {
+        awaitingConfirmation += 1
+        logger.trace(s"Start commit {}. Commits in progress {}", offsets, awaitingConfirmation.toString)
+        consumer.commitAsync(offsets.asJava, new OffsetCommitCallback {
+          override def onComplete(offsets: util.Map[TopicPartition, OffsetAndMetadata], exception: Exception): Unit = {
+            if (exception == null)
+              done.success(Done)
+            else
+              done.failure(exception)
+            logger.trace(s"Commit completed {} - {}", offsets, done.future.value)
+            decrementConfirmation.invoke(())
+          }
+        })
+        poll()
+      }
+
+      override protected def onTimer(timerKey: Any): Unit = {
+        timerKey match {
+          case Poll =>
+            logger.trace("Scheduled poll")
+            pollScheduled = false
+            poll()
+          case msg => super.onTimer(msg)
+        }
+      }
+
+      override def postStop(): Unit = {
+        logger.debug("Stage completed. Closing kafka consumer")
+        consumer.close()
+        stoppedPromise.success(Done)
```

Comment:
I don't think we care, that is communicated downstreams. The Done Future is more a signal for completion of the Control.stop call